internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -0,0 +1,566 @@
+	receipts := []message.Receipt{}
+
+	// update current epoch
+	vm.currentEpoch = epoch
+
+	// create message tracker
+	// Note: the same message could have been included by more than one miner
+	seenMsgs := make(map[cid.Cid]struct{})
+
+	// process messages on each block
+	for _, blk := range blocks {
+		if blk.Miner.Protocol() != address.ID {
+			panic("precond failure: block miner address must be an IDAddress")
+		}
+		// Process block miner's Election PoSt.
+		epostMessage := makeEpostMessage(blk.Miner)
+		if _, err := vm.applyImplicitMessage(epostMessage); err != nil {
+			return nil, err
+		}
+
+		// initial miner penalty
+		// Note: certain msg execution failures can cause the miner to pay for the gas
+		minerPenaltyTotal := types.ZeroAttoFIL
+
+		// Process BLS messages from the block
+		for _, m := range blk.BLSMessages {
+			// do not recompute already seen messages
+			mcid := msgCID(m)
+			if _, found := seenMsgs[mcid]; found {
+				continue
+			}
+
+			// apply message
+			receipt, minerPenaltyCurr := vm.applyMessage(m, m.OnChainLen(), blk.Miner)
+
+			// accumulate result
+			minerPenaltyTotal = minerPenaltyTotal.Add(minerPenaltyCurr)
+			receipts = append(receipts, receipt)
+
+			// flag msg as seen
+			seenMsgs[mcid] = struct{}{}
+		}
+
+		// Process SECP messages from the block
+		for _, sm := range blk.SECPMessages {
+			// extract unsigned message part
+			m := sm.Message
+
+			// do not recompute already seen messages
+			mcid := msgCID(&m)
+			if _, found := seenMsgs[mcid]; found {
+				continue
+			}
+
+			// apply message
+			// Note: the on-chain size for SECP messages is different
+			receipt, minerPenaltyCurr := vm.applyMessage(&m, sm.OnChainLen(), blk.Miner)
+
+			// accumulate result
+			minerPenaltyTotal = minerPenaltyTotal.Add(minerPenaltyCurr)
+			receipts = append(receipts, receipt)
+
+			// flag msg as seen
+			seenMsgs[mcid] = struct{}{}
+		}
+
+		// Pay block reward.
+		rewardMessage := makeBlockRewardMessage(blk.Miner, minerPenaltyTotal)
+		if _, err := vm.applyImplicitMessage(rewardMessage); err != nil {
+			return nil, err
+		}
+	}
+
+	// cron tick
+	firstMiner := blocks[0].Miner
+	cronMessage := makeCronTickMessage(firstMiner)
+	if _, err := vm.applyImplicitMessage(cronMessage); err != nil {
+		return nil, err
+	}
+
+	// commit state
+	// flush all objects out
+	if err := vm.store.Flush(); err != nil {
+		return nil, err
+	}
+	// commit new actor state
+	if err := vm.state.Commit(context.Background()); err != nil {
+		return nil, err
+	}
+	// TODO: update state root (issue: #3718)
+
+	return receipts, nil
+}
+
+// applyImplicitMessage applies messages automatically generated by the vm itself.
+//
+// This messages do not consume client gas are do not fail.
+func (vm *VM) applyImplicitMessage(imsg internalMessage) (out interface{}, err error) {
+	defer func() {
+		if r := recover(); r == nil {
+			switch r.(type) {
+			case runtime.AbortPanicError:
```

Comment:
Can we remove this type?

Reply:
Done, merged the two into:
```go
// ExecutionPanic is used to abort vm execution with an exit code.
type ExecutionPanic struct {
	msg  string
	code exitcode.ExitCode
}
```

And left this two methods: 
```go
func Abort(code exitcode.ExitCode){}

func Abortf(code exitcode.ExitCode, msg string, args ...interface{}){}
```